File: 287.find-the-duplicate-number.cpp
/*
 * @lc app=leetcode id=287 lang=cpp
 *
 * [287] Find the Duplicate Number
 */

// @lc code=start
#include <vector>

class Solution {
public:
    int findDuplicate(std::vector<int>& nums) {
        int slow = 0, fast = 0;
        bool match = false;

        while (!match) {
            slow = nums[slow];
            fast = nums[nums[fast]];

            match = (slow == fast);
        }

        slow = 0;
        if (nums[slow] != nums[fast]) {
            match = false;

            while (!match) {
                slow = nums[slow];
                fast = nums[fast];
                match = (nums[slow] == nums[fast]);
            }
        }

        return nums[slow];
    }
};

// @lc code=end

File: 287.find-the-duplicate-number.py
#
# @lc app=leetcode id=287 lang=python3
#
# [287] Find the Duplicate Number
#

# @lc code=start
from typing import List


class Solution:
    def findDuplicate(self, nums: List[int]) -> int:
        slow, fast = 0, 0
        match = False
        while not match:
            slow = nums[slow]
            fast = nums[fast]
            fast = nums[fast]
            if nums[slow] == nums[fast]:
                match = True
        slow = 0
        if nums[slow] != nums[fast]:
            match = False
            while not match:
                slow = nums[slow]
                fast = nums[fast]
                if nums[slow] == nums[fast]:
                    match = True
        return nums[slow]
# @lc code=end

